When attempting to setup a pseudo name for my g...

Notes
======
When attempting to setup a pseudo name for my group to identify my backup. This is at the backup tab of the using the backup to a friend option.

Environment
======
Environment Key | Value
:----: | :----:
version | 3.3
build | 113
UDID | cf50cfa8152180f7
Android Release | 4.2.2
Android SDK | 17
Brand | Lenovo
Device | A10
Stack trace
======
```java
java.lang.RuntimeException: Parcelable encountered IOException writing serializable object (name = org.apache.harmony.xnet.provider.jsse.OpenSSLRSAPrivateKey)
	at android.os.Parcel.writeSerializable(Parcel.java:1279)
	at android.os.Parcel.writeValue(Parcel.java:1233)
	at android.os.Parcel.writeMapInternal(Parcel.java:591)
	at android.os.Bundle.writeToParcel(Bundle.java:1627)
	at android.os.Parcel.writeBundle(Parcel.java:605)
	at android.content.Intent.writeToParcel(Intent.java:6660)
	at android.app.ActivityManagerProxy.bindService(ActivityManagerNative.java:2729)
	at android.app.ContextImpl.bindService(ContextImpl.java:1431)
	at android.app.ContextImpl.bindService(ContextImpl.java:1407)
	at android.content.ContextWrapper.bindService(ContextWrapper.java:473)
	at com.door43.translationstudio.newui.ShareWithPeerDialog.onStart(ShareWithPeerDialog.java:289)
	at android.app.Fragment.performStart(Fragment.java:1719)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:913)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1057)
	at android.app.BackStackRecord.run(BackStackRecord.java:682)
	at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1435)
	at android.app.FragmentManagerImpl$1.run(FragmentManager.java:441)
	at android.os.Handler.handleCallback(Handler.java:725)
	at android.os.Handler.dispatchMessage(Handler.java:92)
	at android.os.Looper.loop(Looper.java:137)
	at android.app.ActivityThread.main(ActivityThread.java:5041)
	at java.lang.reflect.Method.invokeNative(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:511)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
	at dalvik.system.NativeStart.main(Native Method)
Caused by: java.io.NotSerializableException: org.apache.harmony.xnet.provider.jsse.OpenSSLKey
	at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1364)
	at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
	at java.io.ObjectOutputStream.writeFieldValues(ObjectOutputStream.java:979)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:368)
	at java.io.ObjectOutputStream.writeHierarchy(ObjectOutputStream.java:1074)
	at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1404)
	at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
	at android.os.Parcel.writeSerializable(Parcel.java:1274)
	... 25 more

```
Log history
======
```java
3/29/16 2:27 PM  E/com.door43.translationstudio.newui.ShareWithPeerDialog: Failed to initialize the server service
java.lang.RuntimeException: Parcelable encountered IOException writing serializable object (name = org.apache.harmony.xnet.provider.jsse.OpenSSLRSAPrivateKey)
	at android.os.Parcel.writeSerializable(Parcel.java:1279)
	at android.os.Parcel.writeValue(Parcel.java:1233)
	at android.os.Parcel.writeMapInternal(Parcel.java:591)
	at android.os.Bundle.writeToParcel(Bundle.java:1627)
	at android.os.Parcel.writeBundle(Parcel.java:605)
	at android.content.Intent.writeToParcel(Intent.java:6660)
	at android.app.ActivityManagerProxy.startService(ActivityManagerNative.java:2659)
	at android.app.ContextImpl.startServiceAsUser(ContextImpl.java:1366)
	at android.app.ContextImpl.startService(ContextImpl.java:1353)
	at android.content.ContextWrapper.startService(ContextWrapper.java:450)
	at com.door43.translationstudio.newui.ShareWithPeerDialog.initializeService(ShareWithPeerDialog.java:319)
	at com.door43.translationstudio.newui.ShareWithPeerDialog.onStart(ShareWithPeerDialog.java:283)
	at android.app.Fragment.performStart(Fragment.java:1719)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:913)
	at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1057)
	at android.app.BackStackRecord.run(BackStackRecord.java:682)
	at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1435)
	at android.app.FragmentManagerImpl$1.run(FragmentManager.java:441)
	at android.os.Handler.handleCallback(Handler.java:725)
	at android.os.Handler.dispatchMessage(Handler.java:92)
	at android.os.Looper.loop(Looper.java:137)
	at android.app.ActivityThread.main(ActivityThread.java:5041)
	at java.lang.reflect.Method.invokeNative(Native Method)
	at java.lang.reflect.Method.invoke(Method.java:511)
	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:793)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:560)
	at dalvik.system.NativeStart.main(Native Method)
Caused by: java.io.NotSerializableException: org.apache.harmony.xnet.provider.jsse.OpenSSLKey
	at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1364)
	at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
	at java.io.ObjectOutputStream.writeFieldValues(ObjectOutputStream.java:979)
	at java.io.ObjectOutputStream.defaultWriteObject(ObjectOutputStream.java:368)
	at java.io.ObjectOutputStream.writeHierarchy(ObjectOutputStream.java:1074)
	at java.io.ObjectOutputStream.writeNewObject(ObjectOutputStream.java:1404)
	at java.io.ObjectOutputStream.writeObjectInternal(ObjectOutputStream.java:1671)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1517)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:1481)
	at android.os.Parcel.writeSerializable(Parcel.java:1274)
	... 26 more

     
```

--- a/app/src/main/java/com/door43/translationstudio/newui/ShareWithPeerDialog.java
+++ b/app/src/main/java/com/door43/translationstudio/newui/ShareWithPeerDialog.java
@@ -44,6 +44,8 @@ import org.json.JSONException;
 
 import java.io.File;
 import java.security.InvalidParameterException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
 import java.util.ArrayList;
 import java.util.Locale;
 
@@ -336,8 +338,21 @@ public class ShareWithPeerDialog extends DialogFragment implements ServerService
             RSAEncryption.generateKeys(privateKeyFile, publicKeyFile);
         }
         // TODO: 11/30/2015 we should use a shared interface for setting parameters so we don't have to manage two sets
-        intent.putExtra(ServerService.PARAM_PRIVATE_KEY, RSAEncryption.readPrivateKeyFromFile(privateKeyFile));
-        intent.putExtra(ServerService.PARAM_PUBLIC_KEY, RSAEncryption.getPublicKeyAsString(RSAEncryption.readPublicKeyFromFile(publicKeyFile)));
+        PrivateKey privateKey;
+        PublicKey publicKey;
+        try {
+            privateKey = RSAEncryption.readPrivateKeyFromFile(privateKeyFile);
+            publicKey = RSAEncryption.readPublicKeyFromFile(publicKeyFile);
+        } catch (Exception e) {
+            // try to regenerate the keys if loading fails
+            Logger.w(this.getClass().getName(), "Failed to load the p2p keys. Attempting to regenerate...", e);
+            RSAEncryption.generateKeys(privateKeyFile, publicKeyFile);
+            privateKey = RSAEncryption.readPrivateKeyFromFile(privateKeyFile);
+            publicKey = RSAEncryption.readPublicKeyFromFile(publicKeyFile);
+        }
+
+        intent.putExtra(ServerService.PARAM_PRIVATE_KEY, privateKey);
+        intent.putExtra(ServerService.PARAM_PUBLIC_KEY, RSAEncryption.getPublicKeyAsString(publicKey));
         intent.putExtra(ServerService.PARAM_DEVICE_ALIAS, AppContext.getDeviceNetworkAlias());
         Logger.i(this.getClass().getName(), "Starting service " + intent.getComponent().getClassName());
         getActivity().startService(intent);

--- a/app/src/main/java/com/door43/util/RSAEncryption.java
+++ b/app/src/main/java/com/door43/util/RSAEncryption.java
@@ -63,7 +63,7 @@ public class RSAEncryption {
      * @param exp
      * @throws IOException
      */
-    private static void saveKeys(String fileName,BigInteger mod,BigInteger exp) throws IOException{
+    private static void saveKeys(String fileName,BigInteger mod,BigInteger exp) throws Exception {
         FileOutputStream fos = null;
         ObjectOutputStream oos = null;
 
@@ -92,7 +92,7 @@ public class RSAEncryption {
      * @param data
      * @throws IOException
      */
-    public static byte[] encryptData(String data, PublicKey pubKey) throws IOException {
+    public static byte[] encryptData(String data, PublicKey pubKey) throws Exception {
         byte[] dataToEncrypt = data.getBytes();
         byte[] encryptedData = null;
         try {
@@ -111,7 +111,7 @@ public class RSAEncryption {
      * @param data
      * @throws IOException
      */
-    public static String decryptData(byte[] data, PrivateKey privateKey) throws IOException {
+    public static String decryptData(byte[] data, PrivateKey privateKey) throws Exception {
         try {
 //            PrivateKey privateKey = readPrivateKeyFromFile(privFile.getAbsolutePath());
             Cipher cipher = Cipher.getInstance("RSA");
@@ -169,7 +169,7 @@ public class RSAEncryption {
      * @return PublicKey
      * @throws IOException
      */
-    public static PublicKey readPublicKeyFromFile(File file) throws IOException{
+    public static PublicKey readPublicKeyFromFile(File file) throws Exception{
         FileInputStream fis = null;
         ObjectInputStream ois = null;
         try {
@@ -206,7 +206,7 @@ public class RSAEncryption {
      * @return
      * @throws IOException
      */
-    public static PrivateKey readPrivateKeyFromFile(File file) throws IOException{
+    public static PrivateKey readPrivateKeyFromFile(File file) throws Exception{
         FileInputStream fis = null;
         ObjectInputStream ois = null;
         try {

